### ISSUE TITLE
morphology: engine="convolution" returns the image dtype while unfold/shift return the promoted dtype

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4762]
### What happens

With mixed-dtype operands, the result dtype of every `kornia.morphology` function depends on `engine`. `unfold` and `shift` (the two engines `engine="auto"` picks) compute in `torch.promote_types(tensor, kernel)` (or `tensor` and `structuring_element` when one is given) and return that dtype. `convolution` casts the conv weight and bias to the image's dtype since #4544 (the fix for the #4541 crash) and returns the image's dtype, so a wider kernel or structuring element is rounded to the image dtype first.

```python
import torch
from kornia.morphology import dilation

x = torch.rand(1, 1, 5, 6)
k = torch.ones(3, 3)
for label, img, kern in [("float16 image, float32 kernel", x.half(), k), ("float32 image, float64 kernel", x, k.double())]:
    print(label, [str(dilation(img, kern, engine=e).dtype) for e in ("unfold", "shift", "convolution")])
# float16 image, float32 kernel ['torch.float32', 'torch.float32', 'torch.float16']
# float32 image, float64 kernel ['torch.float64', 'torch.float64', 'torch.float32']
```

Measured on kornia 0.9.0rc1 (`main` at 4b13aae42), torch 2.14.0, CPU. `erosion` behaves the same way, and a float64 `structuring_element` on a float32 image does too (`unfold`/`shift` return float64, `convolution` float32). `opening`, `closing`, `gradient`, `top_hat` and `bottom_hat` inherit it.

### Why it matters

Changing `engine` should not change the result dtype. `_resolve_engine` already relies on promotion: its dtype rule reads `torch.promote_types(tensor.dtype, neighborhood.dtype)` because "a float16 image with a float32 kernel computes, and differentiates, in float32". The convolution engine is the odd one out.

### Expected

All engines return the same dtype. The smallest fix is probably to have the convolution branch cast the padded input, weight and bias to the promoted compute dtype instead of the image dtype. That keeps the #4541 crash fixed and matches `unfold`/`shift`. `engine="auto"` never picks `convolution`, so default calls are unaffected.

Found while reviewing the batch-7 Convention blocks (#4743), whose operand-dtype warning describes the `unfold`/`shift` behaviour and links here for the convolution exception.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by ducha-aiki** [comment]
The dtype split also changes values, not only the result dtype. Because the convolution engine casts the `∓max_val` sentinel to the image's dtype, a `uint8` image wraps it on CPU (-1e4 mod 256 = 240), so a masked cell contributes `(x + 240) mod 256` and can win the max:

```python
import torch
from kornia.morphology import dilation

x = torch.tensor([[10, 20, 30]], dtype=torch.uint8)[None, None]
k = torch.tensor([[1.0, 1.0, 0.0]])
dilation(x, k, border_type="constant", engine="convolution")  # uint8 [240, 250, 30]
dilation(x, k, border_type="constant", engine="unfold")       # float32 [20., 30., 30.]
```

`int8`, `int32` and `int64` images happen to give the right values in this example. MPS rejects every integer image in `conv2d` ("Convolution is supported only for Floating types"), and a `bool` image raises on both CPU and MPS. Measured on torch 2.14.0, CPU and MPS, at `main` 4b13aae42. #4743 pins both the dtype split and the wrap as `test_wart_convolution_engine_keeps_image_dtype_4762`, so the fix should invert that pin.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by ducha-aiki** [comment]
Correction to the addendum above: the `uint8` wrap is not a CPU-wide fact. `engine="convolution"` casts the out-of-range `-max_val` sentinel to `uint8`, and C leaves that conversion undefined, so whether it wraps to `240` or saturates to `0` depends on the platform, the torch release and whether the cell lands in a vectorised chunk.

On macOS arm64 with torch 2.14.0, `torch.full((9,), -1e4).to(torch.uint8)` gives `0` in its first eight cells and `240` in the ninth. torch 2.5.1 gives `240` in all nine. A saturated `0` makes the masked cell a member instead:

```python
import torch
from kornia.morphology import dilation

x = torch.full((1, 1, 3, 3), 10, dtype=torch.uint8); x[..., 0, 0] = 200
k = torch.ones(3, 3); k[2, 2] = 0
dilation(x, k, border_type="constant", engine="convolution")[0, 0, 1, 1]  # 200 on arm64 / torch 2.14.0 (masked cell counted)
dilation(x, k, border_type="constant", engine="unfold")[0, 0, 1, 1]       # 10.
```

The three-element `[10, 20, 30]` example above does wrap on both torch releases on arm64, and the Windows CI legs pass it too. In #4743 (`628563a2b`), `test_wart_convolution_engine_keeps_image_dtype_4762` now replays torch's own cast of the same bias rather than hardcoding `[240, 250, 30]`. The fix here still only has to invert the dtype assertions. The pin at the `bool`-image `convolution` line of `test_wart_integer_and_bool_input_4735` also names this issue.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).

**Comment by aryan32134hello** [comment]
Hi @ducha-aiki — I'd like to take this one. I worked on #4541/#4544 (the crash this issue traces back to), so happy to close the loop.

My understanding of scope, please correct me if I'm off:

1. Change engine="convolution" to cast the padded input, weight, and bias to the promoted compute dtype (torch.promote_types(tensor.dtype, neighborhood.dtype)) instead of the image dtype — matching unfold/shift, rather than the other direction.
2. Fix the integer/bool overflow from casting -max_val into a narrow dtype (the uint8 wrap example above) as part of the same change, since it's the same root cause (casting the sentinel into too narrow a dtype before the comparison/conv happens).
3. Invert test_wart_convolution_engine_keeps_image_dtype_4762 (from #4743) once the behavior changes, and update any other pins naming #4762 per the grep convention in AGENTS.md.

Will open a PR referencing this issue once I have it working and tested across dtypes/devices. Let me know if I'm missing anything before I start.

**Comment by ducha-aiki** [comment]
@aryan32134hello it's yours. We can't assign outside contributors, so this comment is the claim. Two changes to the scope:

1. **Point 1 and point 3 are right.** Make `engine="convolution"` compute in `torch.promote_types(tensor.dtype, neighborhood.dtype)` like `unfold` and `shift`. Then invert `test_wart_convolution_engine_keeps_image_dtype_4762` (`tests/morphology/test_dilation.py`) and rewrite the #4762 note in `kornia/morphology/morphology.py`. Those are the only two in-tree references on `main` (`2fa713d0c`). The result dtype changes for a mixed-dtype call such as a float16 image with a float32 kernel, so the fragment is `.breaking` and says what the engine returned before.
2. **Leave point 2 out.** Integer and `bool` images belong to #4735. The maintainer decision there (2026-09-23) is to **reject** non-float images with a kornia-side error before padding or engine dispatch, not to make the sentinel fit a narrow dtype. Keep this PR to floating-point images.

Also coordinate with #4818 (@NiteeshGowda72). It is open with changes requested and rewrites the padding and bias code of the same convolution path in `morphology.py`. Whichever lands second rebases, so keep your diff narrow and mention #4818 in your PR.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).
